### PR TITLE
Phase 6: Per-Node Adaptive Temperature (Transolver++ Eidetic States)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Transolver's slice routing uses a **global** learnable temperature tensor — one scalar per attention head, shared identically across ALL mesh nodes. This is a known bottleneck: nodes in fundamentally different flow regimes (boundary layer, wake, freestream) compete for slice assignments using the same routing sharpness. High temperature → diffuse routing for all nodes; low temperature → sharp clustering for all nodes. Neither is simultaneously optimal.

**Transolver++ (arxiv:2502.02414, ICML 2025)** addresses this directly with "Node-Adaptive Temperature" (Eidetic States): each node's routing temperature is predicted from its own feature vector via a small linear projection, zero-initialized so training starts from the existing global temperature.

The paper reports 41% drag improvement on DrivAerNet++ and 30.4% lift improvement on AirCraft — both large-scale irregular-mesh aerodynamics tasks directly analogous to TandemFoilSet.

**Why this should help p_tan especially:** The aft foil sits in the disturbed wake of the fore foil. Wake nodes, boundary-layer nodes, and freestream nodes have fundamentally different flow physics and should receive different routing sharpness. p_tan = 29.1 vs p_in = 12.1 — 2.5x worse. Global temperature cannot be simultaneously optimal for all three regimes.

Prior related work in this programme:
- PR #1168 (merged): Learnable scalar offset to temperature for tandem domain — still global across nodes
- PR #1323 (merged): Per-head tandem temperature offset — still global across nodes
- Neither experiment made temperature adaptive at the **node level** — that axis is untested.

## Instructions

The change is localized entirely within `Physics_Attention_Irregular_Mesh` in `train.py` (~15 lines).

**Step 1: Add `--per_node_temp` flag** to argparse:
```python
parser.add_argument('--per_node_temp', action='store_true', default=False,
                    help='Per-node adaptive temperature (Transolver++ Eidetic States)')
```
Add `per_node_temp: bool = False` to the Config dataclass.

**Step 2: Add `temp_proj` to `Physics_Attention_Irregular_Mesh.__init__`:**
```python
if cfg.per_node_temp:
    self.temp_proj = nn.Linear(n_hidden, n_heads, bias=False)
    nn.init.zeros_(self.temp_proj.weight)  # CRITICAL: zero-init → starts as global temp
```

**Step 3: Replace the attention computation in forward** (guard with `if cfg.per_node_temp`):

Current code (approximate):
```python
attn = q_slice @ k_node.permute(0, 1, 3, 2)  # [B, heads, num_slices, N]
attn = attn / self.temperature
attn = F.softmax(attn, dim=-1)
```

New code:
```python
if self.cfg.per_node_temp:
    # x_node: [B, N, n_hidden] — node hidden features going into this attention layer
    tau_delta = self.temp_proj(x_node)             # [B, N, n_heads]
    tau_delta = tau_delta.permute(0, 2, 1)         # [B, n_heads, N]
    tau = self.temperature + tau_delta.unsqueeze(2) # [B, n_heads, 1, N] broadcast
    attn = q_slice @ k_node.permute(0, 1, 3, 2)
    attn = attn / tau
    attn = F.softmax(attn, dim=-1)
else:
    attn = q_slice @ k_node.permute(0, 1, 3, 2)
    attn = attn / self.temperature
    attn = F.softmax(attn, dim=-1)
```

**Note on `x_node`:** The node features before the attention projection. Check how the current code passes features into `Physics_Attention_Irregular_Mesh` — you may need to thread them through to the attention call. The zero-init of `temp_proj.weight` is critical: worst case is identical to current baseline.

**Experiment runs** (run 2 seeds for the primary trial):

```bash
# Trial 1: Per-node adaptive temperature, seed 42
python train.py --agent askeladd \
  --wandb_name "askeladd/per-node-temp-s42" \
  --wandb_group phase6/per-node-temp \
  --per_node_temp --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Trial 2: seed 73
python train.py --agent askeladd \
  --wandb_name "askeladd/per-node-temp-s73" \
  --wandb_group phase6/per-node-temp \
  --per_node_temp --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

W&B group: `phase6/per-node-temp`

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | Baseline |
|--------|----------|
| p_in   | **12.1** |
| p_oodc | **6.6**  |
| p_tan  | **29.1** |
| p_re   | **5.8**  |

Single-model references:
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2105 because Cancelling — per-node adaptive temperature was already tested in PR #1879 (Phase 5) and PR #1793 (Phase 3), both null results. Reassigning askeladd to a different hypothesis..

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Individual Run Metrics

| Config | Seed | val/loss | p_in | p_oodc | p_tan | p_re | W&B Run |
|--------|------|----------|------|--------|-------|------|---------|
| Per-node temp | 42 | 0.4444 | 14.3 | 10.1 | 30.6 | 7.4 | r659cpzj |
| Per-node temp | 73 | 0.5999 | 16.1 | 15.3 | 43.7 | 9.5 | lppg75s0 |
| **Mean** | | **0.5222** | **15.20** | **12.70** | **37.15** | **8.45** | |

### Comparison vs Baseline

| Metric | Baseline (8-seed mean) | Per-node temp (2-seed mean) | Δ |
|--------|----------------------|---------------------------|---|
| val/loss | 0.3829 | 0.5222 | **+36.4%** |
| p_in | 13.03 | 15.20 | **+16.7%** |
| p_oodc | 7.83 | 12.70 | **+62.2%** |
| p_tan | 30.29 | 37.15 | **+22.6%** |
| p_re | 6.45 | 8.45 | **+31.0%** |

### Same-Seed Comparison

| Metric | Baseline s42 | Per-node s42 | Baseline s73 | Per-node s73 |
|--------|-------------|-------------|-------------|-------------|
| p_in | ~12.7 | 14.3 (+12.6%) | ~12.2 | 16.1 (+32.0%) |
| p_oodc | ~8.1 | 10.1 (+24.7%) | ~7.6 | 15.3 (+101.3%) |

### Implementation Details

Added `per_node_temp` flag that creates a zero-initialized `nn.Linear(dim, heads, bias=False)` projection in `Physics_Attention_Irregular_Mesh`. In the forward pass, per-node temperature offsets are computed from the input features `x` (pre-head, shape `[B, N, dim]`), reshaped to `[B, heads, N, 1]`, and added to the global temperature before the slice routing softmax. Temperature is clamped at `min=1e-4`.

### Research Note

The researcher agent found that Transolver++ (arxiv:2502.02414) actually uses a **2-layer MLP with GELU** for the temperature projection (not a single linear), operates on **post-head features** (`dim_head` per head, not full `dim`), and uses GELU activation to ensure non-negative offsets. Our simpler single-linear, zero-init approach allows negative offsets which may contribute to instability — if temperature goes near zero due to large negative offsets, the softmax becomes extremely sharp, causing gradient issues.

### Commands Used

```bash
CUDA_VISIBLE_DEVICES={0,1} python train.py --agent askeladd \
  --wandb_name "askeladd/per-node-temp-s{42,73}" \
  --wandb_group phase6/per-node-temp \
  --per_node_temp --seed {42,73} \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

W&B group: `phase6/per-node-temp`

### What Happened

**Per-node adaptive temperature is a significant negative.** All metrics degraded substantially, with catastrophic results on seed 73 (p_oodc +101%, p_tan +44%). Even the "better" seed 42 shows +24.7% p_oodc degradation.

**Why it failed:**

1. **Training instability.** The single zero-initialized linear projection allows large negative temperature offsets, which can push effective temperature near the clamp floor (1e-4). This creates extremely sharp softmax routing — essentially hard assignment of nodes to slices — which causes gradient issues and training instability. The massive variance between seeds (val/loss 0.444 vs 0.600) confirms instability.

2. **Architecture mismatch.** The Transolver++ paper uses a 2-layer MLP with GELU activation (ensuring non-negative offsets) on post-head features, plus Gumbel noise reparameterization during training. Our simplified implementation lacks these stabilization mechanisms.

3. **Existing temperature mechanisms are already rich.** Our codebase already has global temperature, tandem_temp_offset, and zone_temp — three levels of temperature modulation. The per-node projection may be fighting these existing mechanisms.

### Suggested Follow-ups

- **If revisiting this idea**, implement the full Transolver++ formulation: 2-layer MLP with GELU on post-head features (`dim_head`), which guarantees non-negative temperature offsets. This should resolve the instability issue.
- The concept is sound (spatially-varying routing sharpness), but the implementation needs the stability guarantees from the original paper.
- Consider also adding Gumbel-Softmax reparameterization as in Transolver++, which provides additional training regularization.
